<a href="https://colab.research.google.com/github/ohimjosh/CSC-44800/blob/main/AS03/a3_sample_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [25]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[["Temperature °C", "Mols KCL"]]
y = data["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, test_size=0.1)

In [5]:
X

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [6]:
y

0      6.244743e+05
1      5.779610e+05
2      6.196847e+05
3      1.460449e+06
4      4.325726e+04
           ...     
995    1.545661e+06
996    6.737041e+05
997    3.477543e+05
998    8.684794e+05
999    8.476413e+05
Name: Size nm^3, Length: 1000, dtype: float64

In [7]:
X_train

Temperature °C  Mols KCL
225             886       677
988             250        83
414              25        97
560             673       384
70              370       883
..              ...       ...
826              24        89
157             822        15
624             575       600
443             221       522
524              72       359

[900 rows x 2 columns]

In [8]:
X_test

Temperature °C  Mols KCL
265             300       555
626             529       203
674             174       441
423             896       688
312             760       622
..              ...       ...
316             506        75
257             559       926
496             480       819
195             959       786
299             961       206

[100 rows x 2 columns]

In [9]:
y_train


225    1.223371e+06
988    4.469683e+04
414    5.418829e+03
560    5.291530e+05
70     6.801368e+05
           ...     
826    4.786314e+03
157    3.453043e+04
624    7.071857e+05
443    2.411613e+05
524    5.624231e+04
Name: Size nm^3, Length: 900, dtype: float64

In [10]:
y_test

265    3.454007e+05
626    2.222994e+05
674    1.611126e+05
423    1.257172e+06
312    9.656138e+05
           ...     
316    8.213271e+04
257    1.066475e+06
496    8.111646e+05
195    1.536707e+06
299    4.086765e+05
Name: Size nm^3, Length: 100, dtype: float64

## Part 3. Perform a Linear Regression

In [11]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model

reg.predict(np.array([[300,350]]))

/home/ohimjosh/programming/CSC-44800/AS03/venv/lib/python3.8/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([210402.04682579])

In [12]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
reg.score(X_train, y_train)

0.8607117343929102

The linear regression score of 0.8607117343929102 means that there is a 86.07% accuracy for our trained model. The goal is to get a score of 1.0 which is 100% accuracy of correctness.

In [13]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
reg.coef_

array([ 868.46739437, 1028.30190255])

In [14]:
reg.intercept_

-410043.8373780283

h(x) : y = mx + b

h(x) : $Θ_{0}$ + $Θ_{1}x$

h(x) : 876.43826102 * $x_{1}$ + 1025.93060778 * $x_{2}$ - 416335.79172673094


Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [43]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(reg, X, y, cv= 5)
scores

# Report on their finding and their significance

array([0.83918826, 0.87051239, 0.85871066, 0.87202623, 0.84364641])

In [44]:
print("%0.2f accuracy with a standard deviation of %0.2f" % (scores.mean(), scores.std()))

0.86 accuracy with a standard deviation of 0.01


Looking at the outputs of the cross validation scores they are all around 0.8. The standard deviation is 0.01 meaning that our data is consistent and does not move at drastic intervals. The accuracy of the score is 0.86 which is also good since it is close to 1.0.

## Part 5. Using Polynomial Regression

In [24]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree= 2)
poly.fit_transform(X)
# Report on the metrics and output the resultant equation as you did in Part 3.

array([[1.00000e+00, 4.69000e+02, 6.47000e+02, 2.19961e+05, 3.03443e+05,
        4.18609e+05],
       [1.00000e+00, 4.03000e+02, 6.94000e+02, 1.62409e+05, 2.79682e+05,
        4.81636e+05],
       [1.00000e+00, 3.02000e+02, 9.75000e+02, 9.12040e+04, 2.94450e+05,
        9.50625e+05],
       ...,
       [1.00000e+00, 7.91000e+02, 2.13000e+02, 6.25681e+05, 1.68483e+05,
        4.53690e+04],
       [1.00000e+00, 7.69000e+02, 5.53000e+02, 5.91361e+05, 4.25257e+05,
        3.05809e+05],
       [1.00000e+00, 9.19000e+02, 4.52000e+02, 8.44561e+05, 4.15388e+05,
        2.04304e+05]])

In [34]:
model = Pipeline([('poly', PolynomialFeatures(degree=2)),
                  ('linear', LinearRegression(fit_intercept=False))])
model = model.fit(X_train, y_train)
model.named_steps['linear'].coef_

array([ 1.65430067e-05,  1.20000000e+01, -1.04428721e-07,  1.74900094e-11,
        2.00000000e+00,  2.85714287e-02])